RFCs/0031-Multi-Test-Run-Finalization.md

```diff
@@ -0,0 +1,137 @@
+
+When `Run All Tests` is performed in VS, tests for projects can be executed separately. In this case also combining/merging of data attachments is not performed. Coverage reports are not merged. `Analyze Code Coverage for All Tests` is showing coverage report for only one of the test projects.
+
+# Proposed Changes
+1. Introduce a `IDataCollectorAttachments` interface which can be implemented by Test Platform extensions and provide custom logic to combine/merge attachments. Below interface should be used only for providing logic to combine/merge information from data attachments from independent test executions. This interface should **not be** used for modifying separated data attachment. Test Platform will invoke `HandleDataCollectionAttachmentSets` only if at least 2 data attachments related to Collector (through `GetExtensionUris`) are created by independent test executions.
+
+```
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection
+{
+    /// <summary>
+    /// Interface for data collectors add-ins that choose to handle generated attachments
+    /// </summary>
+    public interface IDataCollectorAttachments
+    {
+        /// <summary>
+        /// Gets the attachments Uris, which are handled by current Collector
+        /// </summary>
+        IEnumerable<Uri> GetExtensionUris();
+
+        /// <summary>
+        /// Indicates whether HandleDataCollectionAttachmentSets is associative (e.g. f([a, b, c]) = f([f([a, b]), c]) = f([a, f([b, c])]))
+        /// </summary>
+        bool IsAssociative  { get; }
+
+        /// <summary>
+        /// Reprocess attachments generated by independent test executions
+        /// </summary>
+        /// <param name="attachments">Attachments to be processed</param>
+        /// <param name="progressReporter">Progress reporter. Accepts integers from 0 to 100</param>
+        /// <param name="logger">Message logger</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Attachments after reprocessing</returns>
+        ICollection<AttachmentSet> HandleDataCollectionAttachmentSets(ICollection<AttachmentSet> attachments, IProgress<int> progressReporter, IMessageLogger logger, CancellationToken cancellationToken);
+    }
+}
+```
+
+Method `GetExtensionUris` should provide all Uris for data attachments which are handled by current collector. Test platform will provide to collector only attachments with such Uris. It's also required that result of method `HandleDataCollectionAttachmentSets` will contain only attachments with such Uris.
+
+`IsAssociative` should indicate if `HandleDataCollectionAttachmentSets` is [associative](https://en.wikipedia.org/wiki/Associative_property).
```

Comment:
I like the idea behind this. However do you think this should be the testplatforms responsibility to check isAssociative for each processor and handle to logic to call collectors accordingly? How about if the test platform passes in a context to the processors about whether this is the last test run in a series or something in between? Something like a DataCollectionProcessorContext passed into HandleDataCollectionAttachments which would contain a `IsLastBatch` bool. 
I'd recommend also sending through runsettings in the context, so processors can choose to operate off of some user provider settings.

Reply:
Having isAssociative on processor interface gives us possibility to create performance improvements which can be required in the future. For example TestPlatform can keep map of uri -> isAssociative for all processors. Such map can be provided back to VsTestConsoleWrapper through messages. Wrapper can perform filtering immediately of all incoming attachments.

Not having such flag will cause that we always need to provide all attachments to all processors. For very large project with hundreds of attachments even with lastBatch false all attachments needs to be propagated to all processors.

Reply:
How about renaming it as `supportsIncrementalProcessing` in that case?

---
In reply to: [446938059](https://github.com/microsoft/vstest-docs/pull/235#discussion_r446938059) [](ancestors = 446938059)

Reply:
I was also thinking of an event handler mechanism like we have for DataCollector interface where processors can subscribe to intermediate attachment processing events or a final attachment processing event and all the TestPlatform would need to do is to invoke the right events. That would make it coherent with the model we have but that ordering of data collectors wouldn't be natural in that model. For ordering I'd expect data collectors to add an attribute specifying a number (with a lower number being the highest order) and the TestPlatform can invoke adapters in that order.

---
In reply to: [447192771](https://github.com/microsoft/vstest-docs/pull/235#discussion_r447192771) [](ancestors = 447192771,446938059)

Reply:
Renamed to: supportsIncrementalProcessing